_inc/lib/core-api/wpcom-endpoints/service-api-keys.php

```diff
@@ -268,6 +284,83 @@ public static function validate_service_api_key_mapbox( $key ) {
 		);
 	}
 
+	/**
+	 * Get the site's own Mapbox API key if set, or the WordPress.com's one otherwise.
+	 *
+	 * @return array An array containing the key (if any) and its source ("site" or "wpcom").
+	 */
+	public static function get_service_api_key_mapbox() {
+		// If the site provides its own Mapbox API key, return it.
+		$service_api_key = Jetpack_Options::get_option( self::key_for_api_service( 'mapbox' ) );
+		if ( $service_api_key ) {
+			return self::format_api_key( $service_api_key );
+		}
+
+		// If the site is not WordPress.com, return an empty API key.
+		$site_id = self::get_wpcom_site_id();
+		if ( ( ! self::is_wpcom() && ! jetpack_is_atomic_site() ) || ! $site_id ) {
+			return self::format_api_key();
+		}
+
+		// If there is a cached token, return it.
+		$transient_key = 'wpcom_mapbox_access_token';
+		$cached_token  = get_transient( $transient_key );
+		if ( $cached_token ) {
+			return self::format_api_key( $cached_token, 'wpcom' );
+		}
+
+		// Otherwise retrieve a WordPress.com token.
+		$request_url = 'https://public-api.wordpress.com/wpcom/v2/sites/' . $site_id . '/mapbox';
+		$response    = wp_remote_get( esc_url_raw( $request_url ) );
+		if ( 200 !== wp_remote_retrieve_response_code( $response ) ) {
+			return self::format_api_key();
+		}
+
+		$response_body             = json_decode( wp_remote_retrieve_body( $response ) );
+		$wpcom_mapbox_access_token = $response_body->wpcom_mapbox_access_token;
```

Comment:
Might want to ensure `$response_body->wpcom_mapbox_access_token` is set and not empty, just to be sure, before storing as a transient.

Reply:
On status 200, the `/mapbox` endpoint always return the access token.
If the status is not 200 (e.g. 403 if the site is not allowed, or 404 if the site ID is not found, etc.), we return before setting the transient.

Reply:
Assuming no one other than us changes the endpoint response--not always a safe assumption! The reason I'm extra cautious when storing transients is the danger of caching bad data and the pain it causes having to clear the cache to fix it, even if the underlying problem that caused it is already resolved.